File: backend/api/routes/student.py
"""
backend/api/routes/student.py

FastAPI router for Student Personalization, Daily Quizzes, Spaced Mistake Drills, and Readiness (Milestone 7).
"""

from typing import Any, Dict, Optional, List
from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from database.db import get_engine, get_session_factory
from database.models import User
from backend.api.routes.auth import get_current_user, get_db
from backend.services.student_service import StudentService

router = APIRouter(prefix="/api/student", tags=["Student Hub & Services"])


# -----------------------------------------------------------------------------
# Schemas
# -----------------------------------------------------------------------------
class OnboardingUpdateRequest(BaseModel):
    target_exam: Optional[str] = None
    target_year: Optional[int] = None
    residency_stage: Optional[str] = None
    medical_college: Optional[str] = None
    primary_speciality: Optional[str] = None


class AnswerSyncItem(BaseModel):
    question_id: str
    selected_answer: Optional[str] = None
    time_spent_seconds: Optional[int] = 0
    client_timestamp: Optional[str] = None


class DraftSyncRequest(BaseModel):
    attempt_id: str
    answers: List[AnswerSyncItem]


# -----------------------------------------------------------------------------
# Endpoints
# -----------------------------------------------------------------------------
@router.get("/taxonomies")
def get_curriculum_taxonomies(db: Session = Depends(get_db)):
    """Returns static/dynamic medical examination and subspecialty taxonomy trees."""
    return StudentService.get_taxonomies(db)


@router.patch("/onboarding")
def update_onboarding(
    req: OnboardingUpdateRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Updates the user's adaptive medical onboarding preferences."""
    try:
        res = StudentService.update_onboarding_profile(
            db=db,
            user_id=current_user.id,
            target_exam=req.target_exam,
            target_year=req.target_year,
            residency_stage=req.residency_stage,
            medical_college=req.medical_college,
            primary_speciality=req.primary_speciality,
        )
        return res
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.get("/daily-quiz")
def get_daily_quiz(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Fetches today's high-yield 5-question daily pathology quiz and updates streak."""
    try:
        return StudentService.get_daily_quiz(db, current_user.id)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.get("/continue-learning")
def get_continue_learning(
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Returns active in-progress assessment attempts and top 3 weak topic recommendations."""
    return StudentService.get_continue_learning(db, current_user.id)


@router.get("/readiness")
def get_exam_readiness(
    target_exam: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Calculates the composite exam readiness index (0–100%)."""
    return StudentService.get_exam_readiness(db, current_user.id, target_exam or current_user.target_exam)


@router.get("/mistakes")
def get_mistake_review(
    topic_id: Optional[str] = None,
    repeated_only: bool = False,
    limit: int = 20,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Returns past mistakes with Robbins textbook citations and 1-click remediation blueprint."""
    return StudentService.get_mistake_review(
        db,
        user_id=current_user.id,
        topic_id=topic_id,
        repeated_only=repeated_only,
        limit=limit,
    )


@router.post("/sync-answers")
def sync_draft_answers(
    req: DraftSyncRequest,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    """Idempotently synchronizes batch draft answers for network resilience."""
    try:
        res = StudentService.sync_draft_answers(
            db=db,
            attempt_id=req.attempt_id,
            user_id=current_user.id,
            answers_payload=[a.model_dump() for a in req.answers],
        )
        return res
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))

File: packages/api-client/src/student.ts
/**
 * packages/api-client/src/student.ts
 *
 * Student Hub, Daily Quiz, Mistake Review & Readiness API client.
 */

import { MedicalApiClient, defaultClient } from './client';
import {
  AnswerSyncItem,
  ContinueLearningResponse,
  DailyQuizResponse,
  ExamReadinessResponse,
  MistakeReviewResponse,
  UserProfile,
} from '@medical/shared';

export class StudentApi {
  constructor(private client: MedicalApiClient = defaultClient) {}

  public async updateOnboarding(payload: {
    target_exam?: string;
    target_year?: number;
    residency_stage?: string;
    medical_college?: string;
    primary_speciality?: string;
  }): Promise<UserProfile> {
    return this.client.request<UserProfile>('/api/student/onboarding', {
      method: 'PATCH',
      body: JSON.stringify(payload),
    });
  }

  public async getDailyQuiz(): Promise<DailyQuizResponse> {
    return this.client.request<DailyQuizResponse>('/api/student/daily-quiz', {
      method: 'GET',
    });
  }

  public async getContinueLearning(): Promise<ContinueLearningResponse> {
    return this.client.request<ContinueLearningResponse>('/api/student/continue-learning', {
      method: 'GET',
    });
  }

  public async getExamReadiness(targetExam?: string): Promise<ExamReadinessResponse> {
    const query = targetExam ? `?target_exam=${encodeURIComponent(targetExam)}` : '';
    return this.client.request<ExamReadinessResponse>(`/api/student/readiness${query}`, {
      method: 'GET',
    });
  }

  public async getMistakes(options?: {
    topic_id?: string;
    repeated_only?: boolean;
    limit?: number;
  }): Promise<MistakeReviewResponse> {
    const params = new URLSearchParams();
    if (options?.topic_id) params.set('topic_id', options.topic_id);
    if (options?.repeated_only) params.set('repeated_only', 'true');
    if (options?.limit) params.set('limit', String(options.limit));
    const qs = params.toString() ? `?${params.toString()}` : '';

    return this.client.request<MistakeReviewResponse>(`/api/student/mistakes${qs}`, {
      method: 'GET',
    });
  }

  public async syncAnswers(attemptId: string, answers: AnswerSyncItem[]): Promise<{ success: boolean; attempt_id: string; synced_count: number }> {
    return this.client.request('/api/student/sync-answers', {
      method: 'POST',
      body: JSON.stringify({ attempt_id: attemptId, answers }),
    });
  }
}

export const studentApi = new StudentApi();
